### PR TITLE
WIP: Command processing

### PR DESCRIPTION
**Pull request by kedars**
#### Problem
Is relevant only after https://github.com/project-chip/connectedhomeip/pull/1468 has been merged

 #### Summary of Changes
Add command processing functions to the Cluster

### CODE REVIEW

File: src/lib/datamodel/ClusterOnOff.h [hunk 1]
```diff
@@ -62,6 +70,44 @@ class ClusterOnOff : public Cluster
...
+        case kOnOffCmdIdOn:
+            HandleCommandOn(cmd);
+            break;
+        case kOnOffCmdIdToggle:
+            HandleCommandToggle(cmd);
+            break;
+        default:
+            /* Unsupported */
+            return CHIP_ERROR_INTERNAL;
+            break;
```
Comment (rwalker-apple) [inline src/lib/datamodel/ClusterOnOff.h:146]: unreachable?

File: src/lib/datamodel/Cluster.h [hunk 2]
```diff
@@ -68,6 +69,18 @@ class Cluster : public Deque<Cluster>
         return mAttrs.Find([attrId](Attribute * item) -> bool { return (item->mAttrId == attrId); });
     }
 
+    /**
+     * @brief
+     *   Handle commands specific to this cluster
+     *
+     * @param cmd the command that needs to be processed
+     */
+    virtual CHIP_ERROR HandleCommands(const Command & cmd)
```
Comment (rwalker-apple) [inline src/lib/datamodel/Cluster.h]: recommend calling this "Command()" or "HandleCommand()"

File: src/lib/datamodel/ClusterOnOff.h [hunk 3]
```diff
@@ -62,6 +70,44 @@ class ClusterOnOff : public Cluster
         AddAttribute(&mOnTime);
         AddAttribute(&mOffWaitTime);
     }
+
+    virtual CHIP_ERROR HandleCommandOff(const Command & cmd)
```
Comment (rwalker-apple) [inline src/lib/datamodel/ClusterOnOff.h:84]: why all this breakout in the base class?  why take a Command parameter if it goes unused?

File: src/lib/datamodel/ClusterOnOff.h [hunk 4]
```diff
@@ -39,12 +39,12 @@ static const uint16_t kAttributeIdOnTime             = 0x4001;
 static const uint16_t kAttributeIdOffWaitTime        = 0x4002;
 
 /* Command IDs */
-static const uint16_t kOnOffCmdIdOff                  = 0x00;
-static const uint16_t kOnOffCmdIdOn                   = 0x01;
-static const uint16_t kOnOffCmdIdToggle               = 0x02;
-static const uint16_t kOnOffCmdIdOffWithEffect        = 0x40;
-static const uint16_t kOnOffCmdIdOffWithRecall        = 0x41;
-static const uint16_t kOnOffCmdIdOnWithTimedOff       = 0x42;
+static const uint16_t kOnOffCmdIdOff            = 0x00;
```
Comment (rwalker-apple) [inline src/lib/datamodel/ClusterOnOff.h:42]: cmdid should be a type, maybe?

File: src/lib/datamodel/ClusterOnOff.h (hunk 1 above)
Comment (kedars) [reply src/lib/datamodel/ClusterOnOff.h]: done

File: src/lib/datamodel/ClusterOnOff.h (hunk 3 above)
Comment (kedars) [reply src/lib/datamodel/ClusterOnOff.h]: So my thought process here is that Clusters will have multiple commands: some of them can be handled within the Cluster itself (like in this case), but for some, the application will have to provide the handler.
 
Then, from an overall consistency perspective, if there is a cluster-specific command, the application can be handle it through `HandleCommand<cmdName>(cmd)`. Applications provide these implementations either to override the default implementation, or implement commands that can't be handled within the SDK anyway.

Thoughts?

Comment (rwalker-apple) [reply src/lib/datamodel/ClusterOnOff.h]: I think the only thing that the base clusters should be providing are those methods required for the base clusters to interact with each other.  So: on-off needs to export a vtable that works for things that level-control needs from on-off.   

This level of decomposition of the command in the base class looks a little like overkill.  Are device manufacturers  really looking for this many callbacks?

Comment (kedars) [reply src/lib/datamodel/ClusterOnOff.h]: It seems like an overkill, particularly in the onoff scenario here and could be eliminated. As we move to other scenarios, would it seem that way, or would most commands will need to be handled in the applications anyway?

Comment (kedars) [reply src/lib/datamodel/ClusterOnOff.h]: Currently, applications provide their functionality by overriding the function, thus adding to the vtable. I think what you are suggesting is to limit the overriding to cluster layering functionality, while for applications they can have their function callbacks that don't rely on a vtable?

File: src/lib/datamodel/ClusterOnOff.h (hunk 4 above)
Comment (kedars) [reply src/lib/datamodel/ClusterOnOff.h:42]: Sure, also for the others (clusterID, attrID), will update those in one of the commits

File: src/lib/datamodel/ClusterOnOff.h (hunk 3 above)
Comment (rwalker-apple) [reply src/lib/datamodel/ClusterOnOff.h]: We only need vtable entries where code in other clusters needs to understand the clusterid<->interface mapping.  To override behavior in an application, those functions can be overridden and dispatched to whatever app code, while dispatching to the superclass for default behavior?

File: src/lib/datamodel/ClusterOnOff.h (hunk 4 above)
Comment (kedars) [reply src/lib/datamodel/ClusterOnOff.h]: done

**Comment by kedars** [comment]
Based on the discussions in the chat channels, closing this for now, until we address the concerns listed there through a proposal first

**Pull request closed without merging**